msteams-platform/sbs-botbuilder-conversation-AI.yml

```diff
@@ -0,0 +1,709 @@
+
+            * Moderator: The `OpenAIModerator` reviews both user's input and the model output to detect harmful speech, such as hate speech, self-harm, and violence.
+
+
+        1. Add the following code to define storage and application:
+
+            ```typescript
+            // Define storage and application
+            const storage = new MemoryStorage();
+            const app = new Application<ApplicationTurnState>({
+                storage,
+                ai: {
+                    planner,
+                    moderator,
+                    promptManager,
+                    prompt: 'chat',
+                    history: {
+                        assistantHistoryType: 'text'
+                    }
+                }
+            });
+            ```
+
+            * The `Application` class replaces the Teams Activity Handler class from Bot Framework SDK. The application class has all the information and bot logic required for an app. You can register actions or activity handlers for the app in this class.
+            * The `MemoryStorage` class is imported from the Bot Builder SDK and has a predefined logic to handle any issues when two or more events exist at the same time so that you don't want to worry about handling storage issues and focus on building your application.
+
+
+            1. Under **Import required bot service**, add `MemoryStorage`.
+
+        1.  Add the following code to define actions:
+
+            ```typescript
+            app.ai.action(AI.FlaggedInputActionName, async (context, state, data) => {
+                await context.sendActivity(`I'm sorry your message was flagged: ${JSON.stringify(data)}`);
+                return false;
+            });
+
+            app.ai.action(AI.FlaggedOutputActionName, async (context, state, data) => {
+                await context.sendActivity(`I'm not allowed to talk about such things.`);
+                return false;
+            });
+            ```
+            
+            When the `OpenAIModerator` reviews and identifies a flag in an input or output, it triggers the action `AI.FlaggedInputActionName` or `AI.FlaggedOutputActionName` and the `context.sendActivity` sends a message to the user. If the user receives a message for an flagged  input or output, the user can control what he wants to do.
+
+        1. Add the following code to define history:
+
+            ```typescript
+            app.message('/history', async (context, state) => {
+            const history = ConversationHistory.toString(state, 2000, '\n\n');
+            await context.sendActivity(history);
+            });
+            ```
```

Comment:
This code does not define the {{$history}} variable. What this code is doing is:

When the user sends '/history' to the bot in Teams, the server will send a log of the conversation history back to the user.

Reply:
The history is actually being tracked under the hood in the AI module and populated in the `state.history` property where `state` is the `ApplicationTurnState`:

![image](https://github.com/MicrosoftDocs/msteams-docs/assets/115390646/d85d78e4-23c7-484d-8753-868359ec04c2)

Reply:
In this prompt template (i.e the `skprompt.txt` file) the {{$history}} gets its value from the `state.history` property. Same for {{$input}}, it's getting the value from `state.input`.

Reply:
Rule of thumb: 
The follow of arbitrary data in the entire bot application is done through the turn state `state` object.

Reply:
Updated the information on the history variable and prompt template.

@singhk97 , Can you please provide more context on "The follow of arbitrary data in the entire bot application is done through the turn state object." and info on the `ApplicationTurnState` object.

Thanks,